src/include/optrack.h

```diff
@@ -0,0 +1,73 @@
+/*-
+ * Copyright (c) 2014-2017 MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+#define	WT_OPTRACK_MAXRECS 16384
+#define	WT_OPTRACK_BUFSIZE WT_OPTRACK_MAXRECS * sizeof(WT_TRACK_RECORD)
+
+/*
+ * WT_TRACK_RECORD --
+ *     A structure for logging potentially long operations.
+ *
+ * We pad the record so that the size of the entire record is
+ * three double words, or 24 bytes. If we don't do this, the compiler
+ * will pad it for us, because we keep records in the record buffer
+ * array and each new record must be aligned on the 8-byte boundary,
+ * since its first element is an 8-byte timestamp. Instead of letting
+ * the compiler insert the padding silently, we pad explicitly, so that
+ * whoever writes the binary decoder can refer to this struct to find
+ * out the record size.
+ */
+struct __wt_track_record {
+	uint64_t timestamp;
+	uint64_t op_id;
+	uint16_t op_type;
+	char padding[6];
+};
+
+#define	WT_TRACK_OP(s, optype)						\
+	tr->timestamp = __wt_rdtsc(s);					\
+	tr->op_type = optype;						\
+	tr->op_id = (uint64_t)&__func__;				\
+									\
+	if (optype == 0 && !id_recorded)				\
+		__wt_optrack_record_funcid(s, tr->op_id,		\
+					   (void*)&__func__,		\
+					   sizeof(__func__),		\
+					   &id_recorded);		\
+	if (s->optrackbuf_ptr == WT_OPTRACK_MAXRECS) {			\
+		if (s->optrack_fh != NULL) {				\
+			size_t bytes = __wt_optrack_flush_buffer(s);	\
+			s->optrack_offset += bytes;			\
+		}                                                       \
+		s->optrackbuf_ptr = 0;					\
+	}
+
+/*
+ * We do not synchronize access to optrack buffer pointer under the assumption
```

Comment:
A session is a thread context in WT. Sessions aren't thread safe, we can't have multiple threads using the same session. This assumption always holds good. Do we need to just update the comment, or is there more to it ?

Reply:
@sulabhM, there were multiple threads using the same session. That's why I had the race. It was a corner case, but it did happen. This assumption may not always hold true.

Reply:
There definitely should not be multiple threads using the same session.  They can use the same session sequentially but not simultaneously.

Reply:
I hope I am wrong. I saw that in the fops test. It was only for session #1 and only at the very end of the execution. So could be the corner case. I will see if I can produce a patch with printfs that reproduce the situation.

Reply:
Ran into issues. Operation tracking was turned off by default on all tests. When I turned it back on in fops test to reproduce the issue, that test blows up due to "WT_SESSION.create: unknown configuration key: 'operation_tracking'". I'll investigate that first. Then I will be set up to try and reproduce the issue with two threads accessing session 1 simultaneously.

Reply:
@sueloverso, first of all, my apologies that I missed your question about threads using the same session that you posted two weeks ago. I just saw it now. Here is some more information on the issue:

1. I have not been able to reproduce it in the latest version of the branch; I will keep trying, because if the issue is there, we need to know about it.

2. I was able to reproduce it in the version of the branch as of commit 95d056a2bb8a3bd8f5fba27aef262e7cfc6c2241.

3. It is visible if we insert some printfs. Here is what I see:

16384, s 0, tid 139660021970688, begin track
16386, s 0, tid 139660136994560, begin track
16386, s 0, tid 139660136994560, end track
16385, s 0, tid 139660162172672, begin track
16387, s 0, tid 139660162172672, end track
16387, s 0, tid 139660136994560, begin track
16388, s 0, tid 139660136994560, end track
16388, s 0, tid 139660162172672, begin track
16388, s 0, tid 139660162172672, end track
16386, s 0, tid 139660021970688, end track

The first number is just an operation tracking buffer pointer; ignore it for the time being, it is being printed in order to help us find the place where the interleaving occurs. The second number is the session id ("s 0" means session 0). The third number is the thread ID obtained via pthread_self(). Threads print the "begin track" message when they enter an operation tracking macro and the "end track" message when they are done with that macro. The macro is just a few lines of non-reentrant code without any synchronization. If there is only one thread using a session at a time, we expect that any thread that prints a "begin track" message in a given session will print the corresponding "end track" message before any other thread using the same session prints the "begin track" message. That is NOT what we see in the snipped above.

What you notice instead is that thread with tid 139660021970688 prints the "begin track", then a bunch of other threads print "begin track" and "end track" messages, and then tid 139660021970688 prints its "end track" message. All of these threads are using session 0, so we see interleaving among multiple threads using the same session.

I observed that behaviour on Alex's AWS instance. To reproduce it:  

1. Check out wt-3133-track-long-ops branch.
2. Check out commit 95d056a2bb8a3bd8f5fba27aef262e7cfc6c2241.
3. Apply the attached patch.
4. Build.
5. Run the fops test: cd build_posix/test/fops; ./t
6. Wait for the test to blow up.
7. Scroll up to search for print statements, where the operation tracking buffer points (the first number printed) reaches 16384. Then look for interleaving among threads.

Note that the attached file is actually a diff. I had to give it the TXT extension, because otherwise GitHub won't upload it. 

I will work on reproducing the bug on the current branch. I won't spend too much time on it, because it is likely that the many merges from develop that happened since the old commit, where I am able to reproduce the bug, eliminated the bug, making this a moot point.

[many-threads-same-session-patch.txt](https://github.com/wiredtiger/wiredtiger/files/1415680/many-threads-same-session-patch.txt)

Reply:
@sueloverso, @sulabhM, bad news. The interleaving does occur in the latest version of the branch. Here is the output from the fops test with additional print statements (see my previous email for instructions on how to interpret the output):

16383, s 0, tid 139730455553792, begin track
16383, s 0, tid 139730184304384, begin track
1, s 0, tid 139730184304384, end track
1, s 0, tid 139730184304384, begin track
2, s 0, tid 139730184304384, end track
0, s 0, tid 139730455553792, end track

The catch here is that we see the 'begin track' message from thread 139730455553792, then a bunch of begin/end messages from thread 139730184304384, then the 'end track' message from thread 139730455553792.

If you apply the attached patch to the current version of the branch (commit baaf96b5cd9a068a35574100d3329fe9e66c7ec5), you should be able to catch the interleaving after a few tries. 

Instructions:

1. Check out wt-3133-track-long-ops branch.
2. Check out commit baaf96b5cd9a068a35574100d3329fe9e66c7ec5 (currently the head).
3. Apply the attached patch.
4. In test/fops/t.c change the operation tracking directory "/mnt/fast/sasha" to any directory you'd like. 
4. Build.
5. Run the fops test: cd build_posix/test/fops; ./t
6. The test will pass.  Look for interleaved begin/end statements among different threads using session 0.


[many-threads-same-session-interleaved.txt](https://github.com/wiredtiger/wiredtiger/files/1416439/many-threads-same-session-interleaved.txt)